### PR TITLE
Gh 2125 new koryphe functions

### PR DESCRIPTION
**Pull request by gchq-21716**

### CODE REVIEW

File: library/time-library/src/main/java/uk/gov/gchq/gaffer/time/function/FilterTimestampSetByTimeRange.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+import java.time.Instant;
+
+/**
+ * A {@code FilterTimestampSetByTimeRange} is a {@link uk.gov.gchq.koryphe.function.KorypheFunction} that takes in a
+ * {@link RBMBackedTimestampSet} and filters the internal timestamps by a start time and end time.
+ */
+@Since("1.9.0")
+@Summary("Filters a RBMBackedTimestampSet by a time range")
+public class FilterTimestampSetByTimeRange extends KorypheFunction<RBMBackedTimestampSet, RBMBackedTimestampSet> {
```
Comment (d47853) [inline library/time-library/src/main/java/uk/gov/gchq/gaffer/time/function/FilterTimestampSetByTimeRange.java]: This should probably be called something like Mask rather than Filter. To me, filter implied you were creating a new object rather than mutating the original object (more like a transform function).

File: library/time-library/src/main/java/uk/gov/gchq/gaffer/time/RBMBackedTimestampSet.java [hunk 2]
```diff
@@ -129,6 +129,18 @@ public Instant getLatest() {
         return getInstantFromInt(it.next());
     }
 
+    /**
+     * Filters by a time range
```
Comment (d47853) [inline library/time-library/src/main/java/uk/gov/gchq/gaffer/time/RBMBackedTimestampSet.java]: Can this be a bit more descriptive?

File: library/time-library/src/test/java/uk/gov/gchq/gaffer/time/RBMBackedTimestampSetTest.java [hunk 3]
```diff
@@ -111,6 +111,26 @@ public void testGetNumberOfTimestamps() {
         assertEquals(4, numberOfTimestamps);
     }
 
+    @Test
+    public void testFilterByTimeRange() {
```
Comment (d47853) [inline library/time-library/src/test/java/uk/gov/gchq/gaffer/time/RBMBackedTimestampSetTest.java]:
```suggestion
    public void shouldFilterByTimeRangeWhenTwoValidTimestampsAreSupplied() {
```

File: library/time-library/src/test/java/uk/gov/gchq/gaffer/time/RBMBackedTimestampSetTest.java [hunk 4]
```diff
@@ -111,6 +111,26 @@ public void testGetNumberOfTimestamps() {
         assertEquals(4, numberOfTimestamps);
     }
```
Comment (d47853) [inline library/time-library/src/test/java/uk/gov/gchq/gaffer/time/RBMBackedTimestampSetTest.java]: Can you add a test for null values in start and end date, as well as for when start date is after end date?

File: library/time-library/src/main/java/uk/gov/gchq/gaffer/time/function/FilterTimestampSetByTimeRange.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    long timeRangeStartEpochMilli;
+    long timeRangeEndEpochMilli;
+
+    public FilterTimestampSetByTimeRange() {
+        // Required for serialisation
+    }
+
+    public FilterTimestampSetByTimeRange(final long timeRangeStartEpochMilli, final long timeRangeEndEpochMilli) {
+        this.timeRangeStartEpochMilli = timeRangeStartEpochMilli;
```
Comment (d47853) [inline library/time-library/src/main/java/uk/gov/gchq/gaffer/time/function/FilterTimestampSetByTimeRange.java]: Should there be support for non-millisecond filtering?

File: library/time-library/src/test/java/uk/gov/gchq/gaffer/time/function/FilterTimestampSetByTimeRangeTest.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
+package uk.gov.gchq.gaffer.time.function;
```
Comment (d47853) [inline library/time-library/src/test/java/uk/gov/gchq/gaffer/time/function/FilterTimestampSetByTimeRangeTest.java]: Can you add a copyright header here

File: library/time-library/src/main/java/uk/gov/gchq/gaffer/time/function/FilterTimestampSetByTimeRange.java (hunk 5 above)
Comment (gchq-21716) [reply library/time-library/src/main/java/uk/gov/gchq/gaffer/time/function/FilterTimestampSetByTimeRange.java]: I hard coded this because it's hard to serialise time unit agnostic objects and simple to serialise longs.

File: library/time-library/src/main/java/uk/gov/gchq/gaffer/time/function/FilterTimestampSetByTimeRange.java (hunk 1 above)
Comment (gchq-21716) [reply library/time-library/src/main/java/uk/gov/gchq/gaffer/time/function/FilterTimestampSetByTimeRange.java]: Good point

**Comment by d47853** [comment]
Looks like you have a checkstyle issue. Run `mvn checkstyle:checkstyle -pl :time-library` to find out what's wrong

File: library/time-library/src/main/java/uk/gov/gchq/gaffer/time/RBMBackedTimestampSet.java [hunk 7]
```diff
@@ -15,6 +15,8 @@
  */
 package uk.gov.gchq.gaffer.time;
 
+import static uk.gov.gchq.gaffer.commonutil.CommonTimeUtil.TimeBucket;
```
Comment (d47853) [inline library/time-library/src/main/java/uk/gov/gchq/gaffer/time/RBMBackedTimestampSet.java]: Undo these changes to the imports to make the checkstyle pass
